modules/network_synchronizer/networked_controller.cpp

```diff
@@ -0,0 +1,1336 @@
+	const Vector<int> peers = get_tree()->get_network_connected_peers();
+	for (int i = 0; i < peers.size(); i += 1) {
+		const int peer_id = peers[i];
+		if (peer_id != get_network_master() && disabled_doll_peers.find(peer_id) == -1) {
+			active_doll_peers.push_back(peer_id);
+		}
+	}
+}
+
+int NetworkedController::calculates_sub_ticks(real_t p_delta, real_t p_iteration_per_seconds) {
+	// Put this to false, so even with sub_ticks = 0 everything will be ok.
+	player_set_has_new_input(false);
+	return controller->calculates_sub_ticks(p_delta, p_iteration_per_seconds);
+}
+
+int NetworkedController::notify_input_checked(uint64_t p_input_id) {
+	return controller->notify_input_checked(p_input_id);
+}
+
+uint64_t NetworkedController::last_known_input() const {
+	return controller->last_known_input();
+}
+
+uint64_t NetworkedController::get_stored_input_id(int p_i) const {
+	return controller->get_stored_input_id(p_i);
+}
+
+bool NetworkedController::process_instant(int p_i, real_t p_delta) {
+	return controller->process_instant(p_i, p_delta);
+}
+
+bool NetworkedController::is_server_controller() const {
+	ERR_FAIL_COND_V(get_tree() == nullptr, false);
+	if (controller_type != CONTROLLER_TYPE_NULL)
+		return controller_type == CONTROLLER_TYPE_SERVER;
+	return get_tree()->is_network_server();
+}
+
+bool NetworkedController::is_player_controller() const {
+	ERR_FAIL_COND_V(get_tree() == nullptr, false);
+	if (controller_type != CONTROLLER_TYPE_NULL)
+		return controller_type == CONTROLLER_TYPE_PLAYER;
+	return get_tree()->is_network_server() == false && is_network_master();
+}
+
+bool NetworkedController::is_doll_controller() const {
+	ERR_FAIL_COND_V(get_tree() == nullptr, false);
+	if (controller_type != CONTROLLER_TYPE_NULL)
+		return controller_type == CONTROLLER_TYPE_DOLL;
+	return get_tree()->is_network_server() == false && is_network_master() == false;
+}
+
+bool NetworkedController::is_nonet_controller() const {
+	ERR_FAIL_COND_V(get_tree() == nullptr, false);
+	if (controller_type != CONTROLLER_TYPE_NULL)
+		return controller_type == CONTROLLER_TYPE_NONET;
+	return get_tree()->get_network_peer().is_null();
+}
+
+void NetworkedController::set_inputs_buffer(const BitArray &p_new_buffer) {
+	inputs_buffer.get_buffer_mut().get_bytes_mut() = p_new_buffer.get_bytes();
+}
+
+void NetworkedController::set_scene_rewinder(SceneRewinder *p_rewinder) {
+	scene_rewinder = p_rewinder;
+}
+
+SceneRewinder *NetworkedController::get_scene_rewinder() const {
+	return scene_rewinder;
+}
+
+bool NetworkedController::has_scene_rewinder() const {
+	return scene_rewinder;
+}
+
+void NetworkedController::_rpc_server_send_inputs(Vector<uint8_t> p_data) {
+	ERR_FAIL_COND(get_tree()->is_network_server() == false);
+
+	const Vector<int> &peers = get_active_doll_peers();
+	for (int i = 0; i < peers.size(); i += 1) {
+		// This is an active doll, Let's send the data.
+		const int peer_id = peers[i];
+		rpc_unreliable_id(peer_id, "_rpc_doll_send_inputs", p_data);
```

Comment:
Rather than blindly replicate from server to dolls and then rewind when there is an issue, it's more common to only send to dolls after the server has validated the input. This way, things like entity interpolation are much easier to implement without visual oddities. Is there a reason you decided to go this unorthodox and more complicated route?

Reply:
It should also be mentioned that broadcasting to every doll every time the server receives input is going to cost much more in terms of network bandwidth. It's much more cost efficient to only broadcast to peers once per server tick (which should be much less than once per frame or even once per physics frame).

Reply:
Of course, when you send every server tick, you still need to consider rewinding in the event of discrepancies. I didn't mean to sound like you didn't need to do that at all.

Reply:
Server tick or Client tick is not so different. Both are processed with the same speed. Right now, there is no reason to wait for the server to validate it, since the client can validate it too. We can avoid virtual lag, and introduce useless complexity, sending it right away.

Later, when doll interpolation will be integrated, that part will disappear.

Reply:
Im suggesting that server tick _should_ be different. Im not talking about the frame rate, im talking about a virtual server tick.

As for i put validation, the whole point is to validate on the server to prevent cheating. The server should be authoritative. That is, nothing the client sends should be trusted and only the server shoukl tell other clients what the world state is. See: 

https://www.gabrielgambetta.com/client-server-game-architecture.html

This is almost always how modern multiplayer games are built in order to mitigate cheating.

Reply:
@AndreaCatania thinking about this more, while an authoritative server model is _usually_ what you want, there are situations (e.g. in certain p2p games) where you don't really need and authoritative server model. I think that having the ability to be server authoritative is important though, as for things like player movement in an MMO, server authority is the  de factor implementation.

In super fast paced games like FPS, lockstep and authoritative servers are not ideal for some things though, so something like this would make more sense, possible couple with other networking concepts like lag compensation or anticheat software.

In the end, I think having the ability to work both ways is important.

I'm not sure what the best way to implement this would be tbh.

Sever authority is actually generally pretty game-specific for the most part and usually is handled via RPCs rather than replications of properties. The one big exception is movement, but perhaps that makes better sense to be implemented separately in a special `NetworkedKinematicBody` node or something that doesn't necessarily use this node.

Reply:
I do think having the ability to configure a logical server tick would be nice though in order to tune bandwidth usage.